### ISSUE TITLE
Add Algolia search to the website

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [typescript-eslint/typescript-eslint#4165]
Splitting out of #4156 as this has been in the works for a while: we need a good docs search solution on the site. Algolia is high quality, free for our kind of open source site, and the de facto standard for Docusaurus.

**Comment by JoshuaKGoldberg** [comment]
Current status: I've started a request with the Algolia folks through their normal intake process. I've told their friendly onboarding person that we do indeed maintain typescript-eslint.io. ⌛

**Comment by JoshuaKGoldberg** [comment]
The Algolia rep emailed back that we should be good to go! https://github.com/algolia/docsearch-configs/blob/master/configs/typescript-eslint.json

Their prefab suggestion (note that we'll use the Docusaurus plugin instead):

```html
<!-- at the end of the HEAD -->
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@docsearch/css@alpha" />

<!-- at the end of the BODY -->
<script type="text/javascript" src="https://cdn.jsdelivr.net/npm/@docsearch/js@alpha"></script>
<script type="text/javascript">
docsearch({
appId: 'BH4D9OD16A',
apiKey: '1ad6b47d4e742c4c0653877b5511c602',
indexName: 'typescript-eslint',
container: '### REPLACE ME ####',
searchParameters: { 'facetFilters': ["type:$TYPE"] },
debug: false // Set debug to true if you want to inspect the modal
});
</script>
```
